dolphin/db/sqlalchemy/api.py

```diff
@@ -309,25 +311,154 @@ def volume_get_all(context, marker=None, limit=None, sort_keys=None,
+    pool_refs = []
+    with session.begin():
+
+        for pool in pools:
+            LOG.debug('adding new pool for original_id {0}:'
+                      .format(pool.get('original_id')))
+            if not pool.get('id'):
+                pool['id'] = uuidutils.generate_uuid()
+
+            pool_ref = models.Pool()
+            pool_ref.update(pool)
+            pool_refs.append(pool_ref)
+
+        session.add_all(pool_refs)
+
+    return pool_refs
+
+
+def pools_delete(context, pools):
+    """Delete multiple pools with the pools dictionary."""
+    session = get_session()
+    pool_refs = []
+    with session.begin():
+        for pool in pools:
+            LOG.debug('deleting pool {0}:'.format(pool.get('id')))
+            query = _pool_get_query(context, session)
+            result = query.filter_by(id=pool.get('id')
+                                     ).delete()
+
+            if not result:
+                LOG.error(exception.PoolNotFound(id=pool.get(
+                    'id')))
+
+            pool_refs.append(pool)
+
+    return pool_refs
 
 
 def pool_update(context, pool_id, values):
     """Update a pool withe the values dictionary."""
-    return NotImplemented
+    session = get_session()
+
+    with session.begin():
+        query = _pool_get_query(context, session)
+        result = query.filter_by(id=pool_id).update(values)
+
+        if not result:
+            LOG.error(exception.PoolNotFound(id=pool_id))
+
+    return result
+
+
+def pools_update(context, pools):
+    """Update multiple pools withe the pools dictionary."""
+    session = get_session()
+
+    with session.begin():
+        pool_refs = []
+
+        for pool in pools:
+            LOG.debug('updating pool {0}:'.format(pool.get('id')))
+            query = _pool_get_query(context, session)
+            result = query.filter_by(id=pool.get('id')
+                                     ).update(pool)
+
+            if not result:
+                LOG.error(exception.PoolNotFound(id=pool.get(
+                    'id')))
+
+            pool_refs.append(pool)
+
+    return pool_refs
 
 
 def pool_get(context, pool_id):
     """Get a pool or raise an exception if it does not exist."""
-    return NotImplemented
+    return _pool_get(context, pool_id)
 
 
 def pool_get_all(context, marker=None, limit=None, sort_keys=None,
                  sort_dirs=None, filters=None, offset=None):
     """Retrieves all storage pools."""
-    return NotImplemented
+    session = get_session()
+    with session.begin():
+        # Generate the query
+        query = _generate_paginate_query(context, session, models.Pool,
+                                         marker, limit, sort_keys, sort_dirs,
+                                         filters, offset,
+                                         )
+        # No pool would match, return empty list
+        if query is None:
+            return []
+        return query.all()
+
+
+def _pool_get_all(context, session=None):
```

Comment:
Seems this function is not used anywhere.

Reply:
done.